TargetSmart 2022: SmartMatch

benstroud opened the pull request:
Adds support for performing TargetSmart SmartMatch list matching workflow executions. SmartMatch matches contact list records to Voterbase, TargetSmart's national database of voting age individuals.

Additional code review of the previously implemented TargetSmart routines has been done to modernize for 2022, including doc updates.

SorenSpicknall commented:
Hey @benstroud, thanks for the contribution! We'll get some eyes on this over the next week or so - I was out of the office last week and our current list of people with review permissions on the repo is relatively small, so we weren't able to immediately dive into these changes.

jburchard commented on docs/targetsmart_automation_workflows.rst:
```diff
@@ -0,0 +1,47 @@
+
+You can then call these methods:
+
+.. code-block:: python
+
+   # Check the status of a workflow execution
+   ts_auto.match_status(job_name='my_job_name')
+
+   # Remove all files for the match job
+   ts_auto.remove_files(job_name='my_job_name')
```
I would consider replacing this with an example of running a match rather than checking the status, as I think that is the method that folks are going to utilize 90% of the time.

```
from parsons import TargetSmartAutomation

match_type = 'standard' # Visit the TargetSmart Automation console for options.
tbl = Table.from_csv('my_file_to_match.csv')
ts_auto = ts_auto.match(tbl, match_type=match_type)
```

jburchard commented on docs/targetsmart_automation_workflows.rst:
```diff
@@ -0,0 +1,47 @@
+workflows that TargetSmart implements for specific client needs. The
+``TargetSmartAutomation`` class can be used to execute these workflows for
+common purposes such as customized list matching workflows.
+
```
I would also include an alert box letting folks know that "For matching applications, TargetSmart SmartMatch is now the recommended solution. See `TargetSmartAPI.smartmatch`."

benstroud replied:
Thanks for the feedback! I'll follow up with a revision as soon as I'm able with other project work.

jburchard commented on parsons/targetsmart/targetsmart_automation.py:
```diff
@@ -9,75 +16,109 @@
 
 # Automation matching documentation can be found here:
-# https://docs.targetsmart.com/developers/automation/index.html
+# https://docs.targetsmart.com/developers/automation/index.html. A custom
+# matching workflow has been implemented by TargetSmart for TMC.
```
Can `TSAutomation` only be used by TMC? Not sure what this note means.

SorenSpicknall replied:
We should certainly get everyone on the same page about this one - if it's just a note in docs about something that TMC does differently than others, it can be struck. If the TargetSmartAutomation class is only useful to TMC, we should probably strike it from the PR and integrate it into TMC work internally. Parsons is widely used both inside and outside TMC, and we try to avoid including anything in the repo that ties it specifically to a TMC use case.

benstroud replied:
Hi all. TargetSmart Automation is a general solution for clients to execute custom workflows that are developed and deployed for specific client needs. Given that, it's a very general interface. For example, the details of the input and output files vary from workflow to workflow. We currently maintain a custom matching workflow for TMC organizations. Other TargetSmart clients also use Automation, but with different custom workflows/layouts/parameters than TMC orgs.

jburchard commented on parsons/targetsmart/targetsmart_automation.py:
```diff
@@ -9,75 +16,109 @@
+                call_back=call_back,
+            )
+            self.sftp.put_file(xml, f"{self.sftp_dir}/{job_name}.job.xml")
+            logger.info("Match configuration uploaded to TargetSmart.")
```
`logger.info(f"Match configuration {job_type} uploaded to TargetSmart.")`

Might be nice to log the `job_type`.

jburchard commented on parsons/targetsmart/targetsmart_smartmatch.py:
```diff
@@ -0,0 +1,311 @@
+        needed.
+
+        `Args:`
+            input_table: Parsons table
+                A Parsons table with `header field names supported by SmartMatch <https://docs.targetsmart.com/developers/tsapis/v2/service/smartmatch.html#supported-field-identifiers>`_. Required.
+            max_matches: int
+                By default only a single best match is returned for an input record. Increase to return additional potentially accurate matches for each input record. Value between 1-10. Default of 1.
+            include_email: bool
+                Set to True to include appended email values for matched records. This is only applicable if your TargetSmart account is configued to return email data. Additional charges may apply if True. Default of False.
+            include_landline: bool
+                Set to True to include appended landline phone number values for matched records. This is only applicable if your TargetSmart account is configued to return landline phone data. Additional charges may apply if True. Default of False.
+            include_wireless: bool
+                Set to True to include appended wireless phone number values for matched records. This is only applicable if your TargetSmart account is configued to return wireless phone data. Additional charges may apply if True. Default of False.
+            include_voip: bool
+                Set to True to include appended VOIP phone number values for matched records. This is only applicable if your TargetSmart account is configued to return VOIP phone data. Additional charges may apply if True. Default of False.
+            tmp_location: str
+                Optionally provide a local directory path where input/output CSV files will be stored. Useful to recover CSV output if downstream ETL processing fails. If not specified, a system tmp location is used. Default of None.
+            keep_smartmatch_input_file: bool
+                Optionally keep the CSV input file that is uploaded in ``tmp_location`` for later use. Default of False.
+            keep_smartmatch_output_gz_file: bool
+                Optionally keep the gzip compressed output file in ``tmp_location`` for later use. The uncompressed output file is always retained in ``tmp_location``. Default of False
```
Could we include some line breaks for readability?

jburchard commented on parsons/targetsmart/targetsmart_smartmatch.py:
```diff
@@ -0,0 +1,311 @@
+
+        if not input_table:
+            raise ValueError(
+                "Missing `input_table`. A Petl table must be provided with"
```
A Parsons Table must be provided.

jburchard commented on parsons/targetsmart/targetsmart_smartmatch.py:
```diff
@@ -0,0 +1,311 @@
+            )
+
+        if not hasattr(input_table, "tocsv"):
+            raise ValueError("`input_table` isn't a valid Petl table.")
```
isn't a valid Parsons table.

jburchard commented on parsons/targetsmart/targetsmart_smartmatch.py:
```diff
@@ -0,0 +1,311 @@
+        TargetSmart system load. SmartMatch executions cannot be canceled once
+        submitted to TargetSmart.
+
+        Since Parsons Petl tables are lazy, the SmartMatch output file is always
```
Since Parsons tables are lazy.

* I think that most users don't know what petl, so just striking for the purposes of clarity.

jburchard commented on parsons/targetsmart/targetsmart_smartmatch.py:
```diff
@@ -0,0 +1,311 @@
+    result join.
+    """
+    if INTERNAL_JOIN_ID in input_table.fieldnames():
+        input_table = input_table.rename(INTERNAL_JOIN_ID, INTERNAL_JOIN_ID_CONFLICT)
```
These are petl methods, but you are accessing a Parsons table. To access the petl object is a accessed via the `table` attribute of a Parsons Table. 

Ex:

`input_table = input_table.table.rename(INTERNAL_JOIN_ID, INTERNAL_JOIN_ID_CONFLICT)`

jburchard commented on parsons/targetsmart/targetsmart_smartmatch.py:
```diff
@@ -0,0 +1,311 @@
+    if INTERNAL_JOIN_ID in input_table.fieldnames():
+        input_table = input_table.rename(INTERNAL_JOIN_ID, INTERNAL_JOIN_ID_CONFLICT)
+
+    return input_table.addrownumbers(field=INTERNAL_JOIN_ID)
```
return `input_table.table.addrownumbers(field=INTERNAL_JOIN_ID)`

jburchard commented on parsons/targetsmart/targetsmart_smartmatch.py:
```diff
@@ -0,0 +1,311 @@
+            f" {submit_filename}."
+        )
+
+        # Write Petl table to CSV and upload for SmartMatch to process
```
Write the Parsons Table.

jburchard commented on parsons/targetsmart/targetsmart_smartmatch.py:
```diff
@@ -0,0 +1,311 @@
+            dir=tmp_location,
+            delete=not keep_smartmatch_input_file,
+        ) as tmp:
+            dataprep_table.tocsv(tmp.name, encoding="utf8")
```
`dataprep_table.tocsv(tmp.name, encoding="utf8")`

jburchard commented on parsons/targetsmart/targetsmart_smartmatch.py:
```diff
@@ -0,0 +1,311 @@
+                    .sort(key=INTERNAL_JOIN_ID)
+                    .cutout(INTERNAL_JOIN_ID)
+                )
+                if INTERNAL_JOIN_ID_CONFLICT in input_table.fieldnames():
```
`if INTERNAL_JOIN_ID_CONFLICT in input_table.columns:`

jburchard commented on parsons/targetsmart/targetsmart_smartmatch.py:
```diff
@@ -0,0 +1,311 @@
+                )
+                outtable = (
+                    petl.leftjoin(  # pylint: disable=no-member
+                        input_table,
```
`input_table.table`

jburchard commented on parsons/targetsmart/targetsmart_smartmatch.py:
```diff
@@ -0,0 +1,311 @@
+                        INTERNAL_JOIN_ID_CONFLICT, INTERNAL_JOIN_ID
+                    )
+
+                return outtable
```
`return Table(outtable)`

Return the output as a Parsons table.

jburchard commented on parsons/targetsmart/targetsmart_smartmatch.py:
```diff
@@ -0,0 +1,311 @@
+import tempfile
+import time
+import uuid
+
```
`from parsons.etl.table import Table`

Using this further down.

jburchard reviewed (changes requested):
This is a really great PR and really well written. Most of my comments fall into two buckets.

1. Minor documentation tweaks and requests for more details.

2. I think that there is a little bit of confusion with petl vs. Parsons tables. Parsons table are basically petl tables and can be accessed as an attribute (`.table`) of a Parsons table. 
 
The `SmartMatch.match()` class takes a Parsons Table, however in various instances, you are calling petl methods on the Parsons Table. I *think* that I identified all of the instances. Alternatively (which might be easier), you can just convert the Parsons table to a petl table at the top of the function and then back at the end.

Awesome work. Thank you!

SorenSpicknall commented:
Hey @benstroud, I wanted to check in on your anticipated timeline for integrating the requested changes. We'd love to get this merged in as soon as you have the space to make your intended modifications.

benstroud commented:
Hello @jburchard @SorenSpicknall. Sincere apologies for the delays. It's been a bit chaotic. The latest commit addresses the PR feedback. Please let me know if you spot additional areas to review. Thanks!
